[flang] Improve intrinsic error messages when multiple signatures exist

tmjbios opened the pull request:
When an intrinsic has multiple table entries (e.g., BESSEL_JN has both elemental and transformational forms), the error message selection logic now prefers specific argument errors over generic "too many actual arguments" messages.

For example, given:

    real, dimension(10) :: xarray
    integer :: n1, n2
    print *, bessel_jn(n1, n2, xarray)

The transformational form of BESSEL_JN(n1, n2, x) requires x to be scalar. Previously, flang reported "too many actual arguments" from the elemental form (which only accepts 2 args), even though the transformational form matched the argument count but failed on rank.

Now flang correctly reports "x= argument has unacceptable rank 1", which better guides the user to the actual problem.

The fix extends the error buffer selection heuristic in Probe() to prefer later table entries when earlier entries fail with argument count mismatches.

tmjbios committed:
[flang] Improve intrinsic error messages when multiple signatures exist

When an intrinsic has multiple table entries (e.g., BESSEL_JN has both
elemental and transformational forms), the error message selection
logic now prefers specific argument errors over generic "too many
actual arguments" messages.

For example, given:

    real, dimension(10) :: xarray
    integer :: n1, n2
    print *, bessel_jn(n1, n2, xarray)

The transformational form of BESSEL_JN(n1, n2, x) requires x to be
scalar. Previously, flang reported "too many actual arguments" from
the elemental form (which only accepts 2 args), even though the
transformational form matched the argument count but failed on rank.

Now flang correctly reports "'x=' argument has unacceptable rank 1",
which better guides the user to the actual problem.

The fix extends the error buffer selection heuristic in Probe() to
prefer later table entries when earlier entries fail with argument
count mismatches.

eugeneepshteyn reviewed (approved):
LGTM, but let's also have the others review

akuhlens reviewed (approved):
Otherwise looks good to me.

tmjbios committed:
Use a more descriptive name for local variable.

The initial name for a local bool wasn't very descriptive.
Update to use preferLaterError as suggested by akuhlens.
